### PR TITLE
[MIG] migration business_requirement from 10.0 to 12.0

### PR DESCRIPTION
**Pull request by ruter-lyu**
@elicoidal @seb-elico

### CODE REVIEW

**Comment by nikul-serpentcs** [comment]
@ruter-lyu pick (preserve commit) commit history also

**Comment by elicoidal** [comment]
@ruter-lyu I have fixed the CLA too: you are covered by Elico eCLA

**Comment by elicoidal** [comment]
check here for the migration advice: https://github.com/OCA/maintainer-tools/wiki

File: business_requirement/README.rst [hunk 1]
```diff
@@ -0,0 +1,347 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
```
Comment (elicoidal) [inline business_requirement/README.rst]: Please switch all modules to LGPL

**Review by elicoidal — COMMENTED** [review]
Not reviewing much: waiting for the commit history to be fixed.
Change the license to LGPL

File: business_requirement/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
+# -*- coding: utf-8 -*-
+# © 2019 Elico Corp (https://www.elico-corp.com).
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+{
+    'name': 'Business Requirement',
+    'category': 'Business Requirements Management',
+    'summary': 'Manage the Business Requirements (stories, scenarios, gaps\
+        and test cases) for your customers',
+    'version': '12.0.1.0.0',
+    'website': 'https://www.elico-corp.com/',
```
Comment (nikul-serpentcs) [inline business_requirement/__manifest__.py]: URL should be `https://github.com/OCA/business-requirement`

File: business_requirement/models/__init__.py [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# -*- coding: utf-8 -*-
```
Comment (nikul-serpentcs) [inline business_requirement/models/__init__.py]: @ruter-lyu no need to `coding` and `copyright` in init file

File: business_requirement/tests/__init__.py [hunk 4]
```diff
@@ -0,0 +1,4 @@
+# -*- coding: utf-8 -*-
```
Comment (nikul-serpentcs) [inline business_requirement/tests/__init__.py]: same as here or other

File: business_requirement/tests/test_br.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
+# -*- coding: utf-8 -*-
+# © 2019 Elico Corp (https://www.elico-corp.com).
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from odoo.tests import common
+from openerp.exceptions import ValidationError
```
Comment (nikul-serpentcs) [inline business_requirement/tests/test_br.py]: odoo

**Review by nikul-serpentcs — CHANGES_REQUESTED** [review]
Improved Code

**Comment by ruter-lyu** [comment]
@elicoidal @seb-elico all fixed, please review again

File: business_requirement/data/business_data.xml [hunk 6]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
```
Comment (nikul-serpentcs) [inline business_requirement/data/business_data.xml]: @ruter-lyu `<odoo noupdate="1">` no need to `<data>`

File: business_requirement/data/business_requirement_demo.xml [hunk 7]
```diff
@@ -0,0 +1,535 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
```
Comment (nikul-serpentcs) [inline business_requirement/data/business_requirement_demo.xml]: same here

File: business_requirement/README.rst [hunk 8]
```diff
@@ -0,0 +1,347 @@
+.. image:: https://img.shields.io/badge/licence-LGPL--3-blue.svg
```
Comment (elicoidal) [inline business_requirement/README.rst]: svg->png

Comment (elicoidal) [reply business_requirement/README.rst]: For the README, can you create fragment similar to https://github.com/OCA/mis-builder/tree/10.0/mis_builder ?

Comment (elicoidal) [reply business_requirement/README.rst]: besides, please update the screenshots to v12

File: business_requirement/data/br_report_paperformat.xml [hunk 9]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <record id="br_report_paperformat_eu" model="report.paperformat">
```
Comment (elicoidal) [inline business_requirement/data/br_report_paperformat.xml:7]: Add Copyright

File: business_requirement/models/business.py [hunk 10]
```diff
@@ -0,0 +1,443 @@
+# © 2019 Elico Corp (https://www.elico-corp.com).
```
Comment (elicoidal) [inline business_requirement/models/business.py]: it should be 2016-2019 everywhere

File: business_requirement/models/business.py [hunk 11]
```diff
@@ -0,0 +1,443 @@
...
+    _inherit = ['mail.thread']
+    _name = "business.requirement"
+    _description = "Business Requirement"
+    _order = 'name desc'
+
+    @api.model
+    def _get_default_company(self):
+        if not self.env.user.company_id:
+            raise ValidationError(
+                _('There is no default company for the current user!'))
```
Comment (elicoidal) [inline business_requirement/models/business.py]: remove "!" (it makes the user feel uncomfortable for no reason)

**Review by elicoidal — CHANGES_REQUESTED** [review]
Not far!

**Comment by elicoidal** [comment]
@seb-elico @joycehuan @JoJoJoJoJoJoJo

File: business_requirement/README.rst (hunk 8 above)
Comment (ruter-lyu) [reply business_requirement/README.rst]: OK, I'll fix it

File: business_requirement/models/business.py [hunk 12]
```diff
@@ -0,0 +1,443 @@
...
+                    }
+                    msg_followers.append((0, 0, msg_vals))
+                if msg_followers:
+                    vals['message_follower_ids'] = msg_followers
+        if vals.get('state'):
+            ir_obj = self.env['ir.model.data']
+            br_xml_id = ir_obj.\
+                get_object('business_requirement',
+                           'group_business_requirement_manager')
+            user = self.env['res.users']
```
Comment (JoJoJoJoJoJoJo) [inline business_requirement/models/business.py]:
```suggestion
            user = self.env.user
```

File: business_requirement/models/business.py [hunk 13]
```diff
@@ -0,0 +1,443 @@
...
+            grps = [grp.id for grp in user.browse(self._uid).groups_id]
+            date = fields.Datetime.now()
+            if vals['state'] == 'confirmed':
+                vals.update({'confirmed_id': user,
```
Comment (JoJoJoJoJoJoJo) [inline business_requirement/models/business.py]:
```suggestion
                vals.update({'confirmed_id': user.id,
```

File: business_requirement/models/business.py [hunk 14]
```diff
@@ -0,0 +1,443 @@
...
+                             'confirmation_date': date})
+            if vals['state'] == 'draft':
+                vals.update({'confirmed_id': False,
+                             'approved_id': False,
+                             'confirmation_date': False,
+                             'approval_date': False
+                             })
+            if vals['state'] == 'approved':
+                if br_xml_id.id in grps:
+                    vals.update({'approved_id': user,
```
Comment (JoJoJoJoJoJoJo) [inline business_requirement/models/business.py]:
```suggestion
                    vals.update({'approved_id': user.id,
```

File: business_requirement/models/business.py [hunk 15]
```diff
@@ -0,0 +1,443 @@
...
+                                 'approval_date': date})
+                else:
+                    raise ValidationError(_(
+                        'You can only move to the following stage: '
+                        'draft/confirmed /cancel/drop.'))
+            if vals['state'] == 'stakeholder_approval':
+                if br_xml_id.id in grps:
+                    vals.update({
+                        'approved_id': user,
```
Comment (JoJoJoJoJoJoJo) [inline business_requirement/models/business.py]:
```suggestion
                        'approved_id': user.id,
```

File: business_requirement/models/business.py [hunk 16]
```diff
@@ -0,0 +1,443 @@
...
+                        'partner_id': project.partner_id.id,
+                        'res_model': self._name,
+                        'subtype_ids': project.subtype_ids.ids
+                    }
+                    msg_followers.append((0, 0, msg_vals))
+                if msg_followers:
+                    vals['message_follower_ids'] = msg_followers
+        if vals.get('state'):
+            ir_obj = self.env['ir.model.data']
+            br_xml_id = ir_obj.\
```
Comment (JoJoJoJoJoJoJo) [inline business_requirement/models/business.py]: Please improve here by `has_group`

File: business_requirement/tests/test_br.py [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+            'project_id': self.pr_1.id
+        }
+        res = self.br.create(br_vals)
+        br_vals1 = {
+            'project_id': self.pr_2.id
+        }
+        res.write(br_vals1)
+        self.assertEqual(res.project_id.id, self.pr_2.id)
+
+    def test_br_read_group(self):
```
Comment (JoJoJoJoJoJoJo) [inline business_requirement/tests/test_br.py:176]: The 2 test functions below seems tricky, finish them or just remove.

**Review by JoJoJoJoJoJoJo — CHANGES_REQUESTED** [review]
Some improvement.

**Comment by ruter-lyu** [comment]
@seb-elico 
All improvement suggestions are done, please review this PR.

By the way, this module is depends by the other modules.

**Comment by elicoidal** [comment]
@ruter-lyu some details to be improved after the functional tests:
- [ ] for the Enterprise Edition, please make sure the icon of the application is the one defined in the module.
- [ ] in demo data, user admin user should be granted "BR manager" and demo user "BR User"
- [ ] In Terms and Conditions demo data: add some "Lorem Ipsum" content
- [ ] Error when printing the Qweb report
```
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/.repo_requirements/odoo/odoo/tools/cache.py", line 88, in lookup
    r = d[key]
  File "/.repo_requirements/odoo/odoo/tools/func.py", line 69, in wrapper
    return func(self, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/tools/lru.py", line 44, in __getitem__
    a = self.d[obj].me
KeyError: ('ir.model.data', <function IrModelData.xmlid_lookup at 0x7f5ebbe422f0>, 'report.html_container')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/.repo_requirements/odoo/odoo/addons/base/models/qweb.py", line 372, in get_template
    document = options.get('load', self.load)(template, options)
  File "/.repo_requirements/odoo/odoo/addons/base/models/ir_qweb.py", line 122, in load
    template = env['ir.ui.view'].read_template(name)
  File "/.repo_requirements/odoo/odoo/addons/base/models/ir_ui_view.py", line 1174, in read_template
    return self._read_template(self.get_view_id(xml_id))
  File "/.repo_requirements/odoo/odoo/addons/base/models/ir_ui_view.py", line 1190, in get_view_id
    return view and view.id or self.env['ir.model.data'].xmlid_to_res_id(template, raise_if_not_found=True)
  File "/.repo_requirements/odoo/odoo/addons/base/models/ir_model.py", line 1389, in xmlid_to_res_id
    return self.xmlid_to_res_model_res_id(xmlid, raise_if_not_found)[1]
  File "/.repo_requirements/odoo/odoo/addons/base/models/ir_model.py", line 1380, in xmlid_to_res_model_res_id
    return self.xmlid_lookup(xmlid)[1:3]
  File "<decorator-gen-25>", line 2, in xmlid_lookup
  File "/.repo_requirements/odoo/odoo/tools/cache.py", line 93, in lookup
    value = d[key] = self.method(*args, **kwargs)
  File "/.repo_requirements/odoo/odoo/addons/base/models/ir_model.py", line 1369, in xmlid_lookup
    raise ValueError('External ID not found in the system: %s' % xmlid)
ValueError: External ID not found in the system: report.html_container

load could not load template
Template: report.html_container
```

**Comment by ruter-lyu** [comment]
> @ruter-lyu some details to be improved after the functional tests:
> 
> * [x]  for the Enterprise Edition, please make sure the icon of the application is the one defined in the module.
> * [x]  in demo data, user admin user should be granted "BR manager" and demo user "BR User"
> * [x]  In Terms and Conditions demo data: add some "Lorem Ipsum" content
> * [x]  Error when printing the Qweb report
> 
> ```
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/.repo_requirements/odoo/odoo/tools/cache.py", line 88, in lookup
>     r = d[key]
>   File "/.repo_requirements/odoo/odoo/tools/func.py", line 69, in wrapper
>     return func(self, *args, **kwargs)
>   File "/.repo_requirements/odoo/odoo/tools/lru.py", line 44, in __getitem__
>     a = self.d[obj].me
> KeyError: ('ir.model.data', <function IrModelData.xmlid_lookup at 0x7f5ebbe422f0>, 'report.html_container')
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/.repo_requirements/odoo/odoo/addons/base/models/qweb.py", line 372, in get_template
>     document = options.get('load', self.load)(template, options)
>   File "/.repo_requirements/odoo/odoo/addons/base/models/ir_qweb.py", line 122, in load
>     template = env['ir.ui.view'].read_template(name)
>   File "/.repo_requirements/odoo/odoo/addons/base/models/ir_ui_view.py", line 1174, in read_template
>     return self._read_template(self.get_view_id(xml_id))
>   File "/.repo_requirements/odoo/odoo/addons/base/models/ir_ui_view.py", line 1190, in get_view_id
>     return view and view.id or self.env['ir.model.data'].xmlid_to_res_id(template, raise_if_not_found=True)
>   File "/.repo_requirements/odoo/odoo/addons/base/models/ir_model.py", line 1389, in xmlid_to_res_id
>     return self.xmlid_to_res_model_res_id(xmlid, raise_if_not_found)[1]
>   File "/.repo_requirements/odoo/odoo/addons/base/models/ir_model.py", line 1380, in xmlid_to_res_model_res_id
>     return self.xmlid_lookup(xmlid)[1:3]
>   File "<decorator-gen-25>", line 2, in xmlid_lookup
>   File "/.repo_requirements/odoo/odoo/tools/cache.py", line 93, in lookup
>     value = d[key] = self.method(*args, **kwargs)
>   File "/.repo_requirements/odoo/odoo/addons/base/models/ir_model.py", line 1369, in xmlid_lookup
>     raise ValueError('External ID not found in the system: %s' % xmlid)
> ValueError: External ID not found in the system: report.html_container
> 
> load could not load template
> Template: report.html_container
> ```

@elicoidal 
Done, pls review again

File: business_requirement/README.rst [hunk 18]
```diff
@@ -0,0 +1,363 @@
+.. |badge1| image:: https://img.shields.io/badge/maturity-Stable-green.png
```
Comment (elicoidal) [inline business_requirement/README.rst]: This should be built by the README generator. You need to fragment the README in pieces similar to https://github.com/OCA/mis-builder/tree/10.0/mis_builder/readme

**Review by elicoidal — APPROVED** [review]
LGTM: we need to work on the readme. Besides it would generate automatically the index.html file

File: business_requirement/README.rst (hunk 18 above)
Comment (ruter-lyu) [reply business_requirement/README.rst]: please review again, is it OK?

**Comment by elicoidal** [comment]
@ruter-lyu I have made some modifications in README and manifest. Please have a look for the other modules.

**Comment by elicoidal** [comment]
@Nikul-Chaudhary @JoJoJoJoJoJoJo your review is welcome.
Functional tests are fine. README is ready

**Comment by elicoidal** [comment]
@ruter-lyu can you try to improve the coverage by adding meaningful unit tests?

**Comment by elicoidal** [comment]
Some improvements for the BRD printout:
* header width is wrong
* Add title "BUSINESS REQUIREMENT DOCUMENT"
* "Change Request" -> "Change Req." (to avoid 2 lines)

Screenshot from 2019-03-19 09-36-46

**Comment by elicoidal** [comment]
@ruter-lyu 
Other points:
- [ ] currently there is a link to Product menu in Configuration. This menu should be in business_requirement_deliverable module not in this menu
- [ ] Stages appears as "false" in the kanban view:

image

**Comment by ruter-lyu** [comment]
> @ruter-lyu
> Other points:
> 
> * [x]  currently there is a link to Product menu in Configuration. This menu should be in business_requirement_deliverable module not in this menu
> * [x]  Stages appears as "false" in the kanban view:
> 
> image

@elicoidal 
Already fixed, please check if there any other problems or not, thanks

br_stage

**Comment by elicoidal** [comment]
When printing a pdf, the document is printed 2 times inside the pdf

**Comment by ruter-lyu** [comment]
> When printing a pdf, the document is printed 2 times inside the pdf

@elicoidal 
fixed.

**Comment by elicoidal** [comment]
@ruter-lyu 
Can you please increase coverage of the UT?

**Comment by elicoidal** [comment]
- [ ] in Kanban view, cannot see the button "Kanban State"
image
- [ ] When drag and drop in KB from Confirmed to stakeholder approval, strange message "You can only move to the following stage: draft/confirmed/cancel/drop.". 
- [ ] Constraint for workflow should be: you can only move to next/previous stage. The message should be improved: in this particular example it should be something like "You can only move to the following stage: Draft/Approved/Cancel/Drop." (Where Approved and Draft will change according to context).
- [ ] Some additional constraints to be added though:
- [ ] From Cancel, we can only go back to Draft
- [ ] From Drop, we can only go back to Draft
- [ ] From Done, we can only go back to Draft
- [ ] I would add the following buttons: Draft/Previous/Next/Cancel/Drop for easier navigation

**Comment by ruter-lyu** [comment]
> * [x]  in Kanban view, cannot see the button "Kanban State"
>   image
> * [x]  Constraint for workflow should be: you can only move to next/previous stage. The message should be improved: in this particular example it should be something like "You can only move to the following stage: Draft/Approved/Cancel/Drop." (Where Approved and Draft will change according to context).
> * [x]  From Cancel, we can only go back to Draft
> * [x]  From Drop, we can only go back to Draft
> * [x]  From Done, we can only go back to Draft

@elicoidal 
please check if it's OK

**Comment by elicoidal** [comment]
The workflow is OK but the kanban state has disappeared from form view and not yet in kanban view: can you add it to both forms? @ruter-lyu

**Comment by ruter-lyu** [comment]
@elicoidal 

Added both kanban view and form view.

image

**Comment by elicoidal** [comment]
Can you move it in KB view to the right along with the photo of the responsible, similar to the tasks:
image

IMO: no need the name of the responsible in KB view (photo is enough) similar as tasks.

**Comment by ruter-lyu** [comment]
@elicoidal 

Already changed.

image

**Comment by elicoidal** [comment]
@ruter-lyu 
I have retested the workflow and changed a bit the usability:
- [ ] From one specific stage you can go to Next stage / Draft / Cancel /Drop (currently is Next stage / Previous / Cancel /Drop)

Besides in the printout:
- [ ] if a html field  (stakeholder story/scenario/gap/use case/T&C) is empty the title is to be removed in the printout.

**Comment by ruter-lyu** [comment]
> @ruter-lyu
> I have retested the workflow and changed a bit the usability:
> 
> * [x]  From one specific stage you can go to Next stage / Draft / Cancel /Drop (currently is Next stage / Previous / Cancel /Drop)
> 
> Besides in the printout:
> 
> * [x]  if a html field  (stakeholder story/scenario/gap/use case/T&C) is empty the title is to be removed in the printout.

@elicoidal 
Done.

**Comment by elicoidal** [comment]
For the workflow, test is OK.
for the printout we have a problem: if I create a new BR and fields are blank, the printout does not show  title.
Nevertheless if I have an existing BR and remove the field, the title still shows, probably for some remaining html blank code:
image

Not sure whether you have a solution for that. @ruter-lyu

**Comment by ruter-lyu** [comment]
> For the workflow, test is OK.
> for the printout we have a problem: if I create a new BR and fields are blank, the printout does not show title.
> Nevertheless if I have an existing BR and remove the field, the title still shows, probably for some remaining html blank code:
> image
> 
> Not sure whether you have a solution for that. @ruter-lyu

@elicoidal 

Already fixed it.

It's because of the non-breaking space `\xa0` in Latin1 (ISO 8859-1)

**Comment by elicoidal** [comment]
> Already fixed it.

seems not: please check runbot @ruter-lyu

**Comment by ruter-lyu** [comment]
> > Already fixed it.
> 
> seems not: please check runbot @ruter-lyu

Done.

**Comment by elicoidal** [comment]
LGTM

**Comment by elicoidal** [comment]
@JoJoJoJoJoJoJo @Nikul-Chaudhary @joycehuan @seb-elico please review

File: business_requirement/models/__init__.py [hunk 19]
```diff
@@ -0,0 +1,3 @@
+from . import business
+from . import project# © 2016-2019 Elico Corp (https://www.elico-corp.com).
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
```
Comment (JoJoJoJoJoJoJo) [inline business_requirement/models/__init__.py:3]: No need for license line in `__init__.py`

File: business_requirement/models/business.py [hunk 20]
```diff
@@ -0,0 +1,483 @@
...
+
+    @api.model
+    def _clean_html_field(self, vals):
+        """
+        Clean the empty HTML field to avoid it be printed
+        :param vals: create or write vals
+        :return:
+        """
+        re_tag = re.compile(r'</?\w+[^>]*>')
+        field_list = ['business_requirement', 'scenario', 'gap',
```
Comment (JoJoJoJoJoJoJo) [inline business_requirement/models/business.py:334]: try not use hard coded field names but get html fields from `self._fields`

**Review by JoJoJoJoJoJoJo — APPROVED** [review]
Small details and LGTM.

File: business_requirement/models/__init__.py (hunk 19 above)
Comment (nikul-serpentcs) [reply business_requirement/models/__init__.py:3]: @JoJoJoJoJoJoJo Keep license, no need to Copyright

**Comment by pedrobaeza** [comment]
Superseeded by #315

**Pull request closed without merging**